Introduced registry-based system for complete automation of validator registration

sivchari opened the pull request.

sivchari commented on cmd/generate-validators/main.go:
```diff
@@ -0,0 +1,298 @@
+	registryFile = "internal/analyzers/govalid/registry_init.go"
+	markersFile  = "internal/markers/markers_generated.go"
+)
+
+var funcMap = template.FuncMap{
+	"firstLetter": func(s string) string {
+		if len(s) > 0 {
+			return strings.ToLower(s[:1])
+		}
+
+		return "x"
+	},
+	"title": cases.Title,
+}
```
Actually, it's unlikely to use multi-byte literal, but it's good to avoid risks, then let's use utf8 package. Thanks.

ccoVeille commented on cmd/generate-validators/main.go:
```diff
@@ -0,0 +1,298 @@
+	}
+
+	// Create output directory if it doesn't exist
+	if err := os.MkdirAll(outputDir, 0750); err != nil {
```
There are still old octal notation, but maybe you will handle it in another PR

sivchari commented:
Hi, @taua-almeida @BlackBuck

I create this PR to contribute more easily and avoid conflict for each rule.
Although the PR is too big, if you have time, please review this.

Before:
  - Manually edit 5-6 files when creating new validators
  - Risk of runtime errors from registration omissions
  - Scattered configuration management

After:
  - Single command make generate-validator MARKER=name auto-generates all files
  - Registry automatically discovers and registers validators
  - Template-driven consistent quality

taua-almeida commented on cmd/generate-validators/main.go:
```diff
@@ -0,0 +1,303 @@
+	}
+
+	// Otherwise, generate all files from existing validators
+	if err := generateAll(); err != nil {
```
this could be dangerous, have you considered delegating scaffold and generate into separate functions in another package? keeping main to flag parsing & dispatch.

taua-almeida commented on cmd/generate-validators/main.go:
```diff
@@ -0,0 +1,303 @@
+	}
+}
+
+func scaffoldValidator(markerName string) error {
```
Right now scaffoldValidator will happily overwrite an existing marker file—it doesn’t check for existing markers.

BlackBuck commented on Makefile:
It would be really helpful if you could include formatting tests inside  the `makefile` itself (it's better to have the contributor test a lot of things on their end than have it in a workflow that requires manual approval by the maintainer; that way, a lot of burden is reduced on the maintainers).

You could add the following lines:

```
.PHONY: fmt
fmt: ## Format Go code and fix imports
    @echo "Formatting Go code..."
    go fmt ./...
    @echo "Fixing imports..."
    go run golang.org/x/tools/cmd/goimports -w .
```

BlackBuck commented on CONTRIBUTING.md:
Additionally, it is usually advised to have a quality checklist for the contributors for them to make sure their code doesn't break when a GitHub workflow runs. Maybe we could add that as well?

sivchari commented on cmd/generate-validators/main.go:
```diff
@@ -0,0 +1,303 @@
+	}
+
+	// Create validator rule file
+	validatorPath := filepath.Join(rulesDir, markerLower+".go")
```
Good catch!!

sivchari commented on internal/analyzers/govalid/templates/validation.go.tmpl:
```diff
@@ -0,0 +1,53 @@
+		{{ if and (ne $parentVariable "") ( .Validators ) }}
+			{{ "}" }}
+			{{ $parentVariable = "" }}
+		{{ end -}}
```
Nice! I use template's block syntax.

sivchari commented:
@taua-almeida @BlackBuck 
Thank you for reviewing. Your feedback is really helpful.
I addressed all review findings, so if you have time, please take a look 😄

taua-almeida commented on cmd/generate-validators/internal/generate/discovery.go:
```diff
@@ -0,0 +1,96 @@
+
+		info, err := analyzeValidatorFile(file)
+		if err != nil {
+			log.Printf("Warning: failed to analyze %s: %v", file, err)
+			continue
```
what is the issue of failing in analyze? I mean this could even happen? If there is a problem in not being able to analyze a file I would consider either a `log.Fatalf` or stop it. Specially for first versions could avoid potential hidden issues

sivchari replied:
Okay, I use Fatal to report it.

sivchari commented on cmd/generate-validators/internal/generate/discovery.go:
```diff
@@ -0,0 +1,96 @@
+		MarkerName:   markerName,
+		FunctionName: validateFunc,
+	}, nil
+}
```
I just thought golangci-lint fmt covers it, but it seems not to work. Thanks.

sivchari replied:
I just thought golangci-lint fmt covers it, but it seems not to work. Thanks.

sivchari commented:
@ccoVeille @taua-almeida 
Hey ho. I addressed findings again. PTAL 🙏
